add hyperlink column to cases and forms in exports

nickpell opened the pull request:
Product Note: Adds a column to exports containing a hyperlink to the case or form represented in that row.

By default, the column is selected in all new exports.  The column is unselected by default in all existing exports.

https://trello.com/c/rxR9cvq6/86-crs-item-for-q4-new-column-in-form-and-case-exports

🐡

Commit:
add case and form links to exports

Commit:
[refactor] turn classmethods into instancemethods

Commit:
form and case links in existing exports should be disabled by default

Commit:
use hyperlink formatting for form/case links in exports

Commit:
use forked openpyxl

Commit:
fix corehq.apps.app_manager.tests.test_bulk_ui_translation:BulkUiTranslation.test_no_change

Commit:
fix corehq.apps.export.tests.test_export_instance:TestCaseExportInstanceGeneration.test_generate_instance_from_schema

millerdev commented on corehq/ex-submodules/couchexport/writers.py:
```diff
@@ -405,10 +405,14 @@ def get_write_value(value):
         if self.format_as_text:
             for cell in cells:
                 cell.number_format = numbers.FORMAT_TEXT
+        for hyperlink_column_index in hyperlink_column_indices:
```
Looks like `hyperlink_column_indices` could be `None`, which would cause an error. Could change the default from `None` to `()` (empty tuple).

millerdev commented on corehq/apps/export/export.py:
```diff
@@ -105,14 +105,14 @@ def open(self, export_instances):
                 self.writer.close()
 
-    def write(self, table, row):
+    def write(self, table, row, hyperlink_column_indices):
```
Adding this argument to every `write()` method bloats the function signature and requires modifications to a lot of code that simply passes it through and shouldn't need to know about it. It doesn't seem like the right level of abstraction—we would not want to add an argument like this for every different formatting option. Ideally this information be included in the `table` object. Would that be possible?

nickpell replied:
Good point.  I updated the code to pass in the column indices with ```ExportRow```, which seems like the best place to do it.  I didn't use the table config in this spot because not all rows get the hyperlink formatting (ie header rows)

Commit:
fix corehq.apps.translations.tests.test_bulk_app_translation

Commit:
fix corehq.apps.app_manager.tests.test_bulk_ui_translation

Commit:
pass hyperlink columns in ExportRow/FormattedRow

Commit:
lint

Commit:
fix corehq.util.workbook_reading.tests.test_spreadsheets

nickpell commented on corehq/util/workbook_reading/tests/test_spreadsheets.py:
```diff
@@ -40,27 +40,53 @@ def _assert_worksheets_equal(self, worksheet1, worksheet2):
+                            ['Date', date(1988, 7, 7)],
+                            ['Date Time', datetime(2016, 1, 1, 12, 0)],
+                            ['Time', time(12, 0)],
+                            ['Midnight', date(1899, 12, 30)],
```
This is strange.  Adding this commit to run remainder of tests, but I want to understand what's going on here before merging.

nickpell replied:
https://bitbucket.org/openpyxl/openpyxl/issues/936/datetime-1899-12-30-0-0-is-returned-when-0

nickpell replied:
Here is the test that had to be fixed:
https://travis-ci.org/dimagi/commcare-hq/jobs/470126342

nickpell replied:
@dannyroberts Any chance you remember how you picked these test cases?  It looks like with the openpyxl upgrade there is a slight difference in how this row is handled:
<img width="141" alt="screen shot 2018-12-19 at 3 23 29 pm" src="https://user-images.githubusercontent.com/1757035/50246009-15fd0700-03a2-11e9-9c72-e9a0416adcdd.png">

Commit:
fix corehq.apps.export.tests.test_sms_export.TestSmsExport

Commit:
fix corehq.apps.userreports.tests.test_report_builder:DataSourceReferenceTest.test_reference_for_forms

Commit:
fix corehq.apps.fixtures.tests.test_upload:TestFixtureUpload.test_row_addition

millerdev commented on corehq/apps/app_manager/tests/test_bulk_ui_translation.py:
```diff
@@ -35,7 +36,12 @@ def test_no_change(self):
         headers = (('translations', ('property', 'en')),)
         f = self._build_translation_download_file(headers)
```
You could make this function return a real file and avoid all other changes to this file.

millerdev commented on corehq/apps/translations/tests/test_bulk_app_translation.py:
```diff
@@ -570,11 +573,13 @@ def setUpClass(cls):
         export_raw(cls.excel_headers, cls.excel_data, file, format=Format.XLS_2007)
 
-        with tempfile.TemporaryFile(suffix='.xlsx') as f:
-            f.write(file.getvalue())
```
I think `f.seek(0)` after this line would make this work.

millerdev commented on corehq/apps/translations/tests/test_bulk_app_translation.py:
```diff
@@ -664,10 +669,11 @@ class AggregateMarkdownNodeTests(SimpleTestCase, TestXmlMixin):
+            with open(temp_path, 'wb') as temp_file:
+                export_raw(self.headers, self.data, temp_file, format=Format.XLS_2007)
+            with open(temp_path, 'rb') as temp_file:
+                workbook = WorkbookJSONReader(temp_file)  # __init__ will read string_io
```
obsolete comment

millerdev reviewed (commented):
Made a bunch of comments on new temporary file handling, but it might be better to fix the deeper issue. It looks like `WorkbookJSONReader` no longer reads from in-memory file types like `BytesIO`, which is unfortunate. Do you think we could fix that root cause of the issue rather than treating the symptoms in various places?

Commit:
Revert "fix corehq.apps.fixtures.tests.test_upload:TestFixtureUpload.test_row_addition"

This reverts commit ff642694f729443e7ab44e72273196967a5f69ad.

Commit:
fix corehq.apps.fixtures.tests.test_upload:TestFixtureUpload.test_row_addition

Commit:
Revert "fix corehq.apps.app_manager.tests.test_bulk_ui_translation"

This reverts commit af50865151099de352abe5450126f7a7259d2795.

Commit:
support passing BytesIO to WorkbookJSONReader

Commit:
Revert "fix corehq.apps.translations.tests.test_bulk_app_translation"

This reverts commit c2827301b10bebff51a432fca085a5e38a2f9a69.

Commit:
fix corehq.apps.translations.tests.test_bulk_app_translation

nickpell commented:
@millerdev I updated ```WorkbookJSONReader``` to handle additional types of objects, and reverted most of the changes to the tests that I had made in this PR.  I also added ```f.seek(0)``` in a few places as necessary.

nickpell commented on corehq/util/workbook_json/excel.py:
```diff
@@ -190,7 +190,7 @@ def _convert_float(value):
 
     def __init__(self, file_or_filename):
-        if isinstance(file_or_filename, InMemoryUploadedFile):
+        if isinstance(file_or_filename, (InMemoryUploadedFile, file)):
```
This is necessary because the name the tempfile name ```file_or_filename.name``` is ```<fdopen>```, which openpyxl is unable to handle.

millerdev commented on corehq/util/workbook_json/excel.py:
```diff
@@ -190,9 +192,13 @@ def _convert_float(value):
 
     def __init__(self, file_or_filename):
-        if isinstance(file_or_filename, InMemoryUploadedFile):
+        if isinstance(file_or_filename, (InMemoryUploadedFile, file, BytesIO)):
```
Adding `file` here makes me nervous. It appears that whatever this was doing in the past worked with bad seek positions, which means either openpyxl was calling `.seek(0)` on the file before reading, or it was using some other method of getting at the content such as `open(file_or_filename.name, ...)`. The way this is reading the file now, `file_or_filename.read()` will return less than the whole file if the seek position is anywhere but at position 0.

Also, I think the way to fix the stickler error is to use `types.FileType`

Commit:
updates to WorkbookJSONReader.__init__

nickpell commented on corehq/util/workbook_json/excel.py:
```diff
@@ -190,9 +192,13 @@ def _convert_float(value):
+            if isinstance(file_or_filename, (InMemoryUploadedFile, file)):
+                value = file_or_filename.read()
+            else:
+                value = file_or_filename.getvalue()
```
@millerdev If I remove ```FileType``` then I get the following error:

```
ERROR: corehq.apps.fixtures.tests.test_upload:TestFixtureUpload.test_row_addition
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nick/Desktop/commcare-hq/corehq/apps/fixtures/tests/test_upload.py", line 215, in test_row_addition
    workbook = self._get_workbook_from_data(self.headers, rows)
  File "/Users/nick/Desktop/commcare-hq/corehq/apps/fixtures/tests/test_upload.py", line 193, in _get_workbook_from_data
    return get_workbook(f)
  File "/Users/nick/Desktop/commcare-hq/corehq/apps/fixtures/upload/workbook.py", line 15, in get_workbook
    return _FixtureWorkbook(file_or_filename)
  File "/Users/nick/Desktop/commcare-hq/corehq/apps/fixtures/upload/workbook.py", line 25, in __init__
    self.workbook = WorkbookJSONReader(file_or_filename)
  File "/Users/nick/Desktop/commcare-hq/corehq/util/workbook_json/excel.py", line 203, in __init__
    self.wb = openpyxl.load_workbook(file_or_filename, read_only=True, data_only=True)
  File "/Users/nick/Desktop/commcare-hq/src/openpyxl-hyperlink/openpyxl/reader/excel.py", line 311, in load_workbook
  File "/Users/nick/Desktop/commcare-hq/src/openpyxl-hyperlink/openpyxl/reader/excel.py", line 270, in read
  File "/Users/nick/Desktop/commcare-hq/src/openpyxl-hyperlink/openpyxl/reader/excel.py", line 167, in read_workbook
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/zipfile.py", line 756, in __init__
    self.fp = open(file, modeDict[mode])
IOError: [Errno 2] No such file or directory: '<fdopen>'
```

'<fdopen>' is the name of the name.

nickpell replied:
The other workaround I used was to change the test code as follows:
https://github.com/dimagi/commcare-hq/pull/22774/commits/ff642694f729443e7ab44e72273196967a5f69ad

millerdev commented on corehq/util/workbook_json/excel.py:
```diff
@@ -1,7 +1,10 @@
+from io import BytesIO
 from zipfile import BadZipfile
 from tempfile import NamedTemporaryFile
+from types import FileType
```
huh, guess I was wrong. Not sure of the right way to do this
